Render optical telescopes

jeisenbe opened the pull request:
Fixes #3389

Changes proposed in this pull request:

- Add "radar" and "gamma" as types of non-optical telescopes, rendered the same as radio telescopes

- Render all other telescopes as optical telescopes (i.e. telescope:type is not radio, gamma or radar), with initial zoom level 17, increasing to 16, 15 and 14 for telescopes greater than 2, 4 or 8 meters in diameter (aperture). Name label rendered at zoom 17.

- Add telescope dome svg icon, courtesy of @Tomasz-W

This will complete the rendering of all man_made=telescopes.

Test rendering with links to the example places:

**Mauna Kea, Hawaii**
Some of the world's biggest telescopes. Only about 100 telescopes are bigger than 2 meters, and only a dozen are bigger than 8 meters and would render at z14. Several are here:
https://www.openstreetmap.org/#map=17/19.82366/-155.47612
z14
![maunakeaz14](https://user-images.githubusercontent.com/42757252/47919257-47a13800-def2-11e8-9962-7d3edee995b6.png)

z15
![maunakeaz15](https://user-images.githubusercontent.com/42757252/47919829-c3e84b00-def3-11e8-8421-844848a9280d.png)

z16
![maunakeaz16](https://user-images.githubusercontent.com/42757252/47919515-e3cb3f00-def2-11e8-8aed-40aa494b83cf.png)

z17
![mauna-kea-17](https://user-images.githubusercontent.com/42757252/47919518-e4fc6c00-def2-11e8-9585-80237468c3e3.png)


**Paranal Observatory, Chile**
The other place with multiple huge telescopes >8 meters (the 4 shown are all part of the VLT; "Very Large Telescope"
https://www.openstreetmap.org/#map=15/-24.6294/-70.4018
z14
![chile-z14](https://user-images.githubusercontent.com/42757252/47919835-c9459580-def3-11e8-9dc1-24dbf2083970.png)

z15
![chilez15](https://user-images.githubusercontent.com/42757252/47920545-a3b98b80-def5-11e8-9579-de59e6447212.png)

z16
![chilez16](https://user-images.githubusercontent.com/42757252/47919912-03169c00-def4-11e8-83c1-cb0f875936a3.png)

z17
![chilez17](https://user-images.githubusercontent.com/42757252/47919921-06aa2300-def4-11e8-9b6e-2f051f31e568.png)

z18
![chilez18](https://user-images.githubusercontent.com/42757252/47920543-a1efc800-def5-11e8-8d6e-8633a924e1a7.png)


**Hampstead Observatory in urban London, England**
An example of a historic/educational telescope in a city. Renders at z17.
https://www.openstreetmap.org/#map=17/52.166756/0.034975
z17
![hampsteadz17](https://user-images.githubusercontent.com/42757252/47920549-a5834f00-def5-11e8-861c-9a71a4a3e08e.png)

z18
![hampsteadz18](https://user-images.githubusercontent.com/42757252/47920555-aa480300-def5-11e8-9faf-e29af2b01f66.png)


**Suburban Berlin Observatory**
Several optical telescopes at a research site on the outskirts of Berlin
https://www.openstreetmap.org/#map=17/52.4049489/13.1030466
z17
![berlinz17](https://user-images.githubusercontent.com/42757252/47920593-c5b30e00-def5-11e8-96aa-55253e5189c4.png)

z18
![berlinz18](https://user-images.githubusercontent.com/42757252/47920589-c3e94a80-def5-11e8-8af1-5ae89cbd8632.png)

kocio-pl commented:
> Add "radar" and "gamma" as types of non-optical telescopes, rendered the same as radio telescopes

Gamma is directly discouraged on the `man_made=telescope` wiki page ("use telescope:type=radio + telescope:spectrum=*") and radar falls into the same category. Their usage is very small (15 and 6 respectively), so we can retag them all in a short time and no special rendering support is needed.

jeisenbe commented:
The wiki is wrong about gamma ray observatories; gamma is shorter
wavelength and higher energy than X-rays, so it’s the opposite of radio
waves, which are very long wavelengths and very low energy per photon.

But I understand about the low usage.

Would it be ok to still exclude them from rendering with the optical
telescope dome icon? Or should I remove that line too?

On Sat, Nov 3, 2018 at 5:52 PM kocio-pl <notifications@github.com> wrote:

> Add "radar" and "gamma" as types of non-optical telescopes, rendered the
> same as radio telescopes
>
> Gamma is directly discouraged on the man_made=telescope wiki page ("use
> telescope:type=radio + telescope:spectrum=*") and radar falls into the same
> category. Their usage is very small (15 and 6 respectively), so we can
> retag them all in a short time and no special rendering support is needed.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gravitystorm/openstreetmap-carto/pull/3492#issuecomment-435571731>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AoxshOTktD-hh28KsqdCG3VIn2TLpjRlks5urVk7gaJpZM4YLvuQ>
> .
>

kocio-pl commented:
It would be good to expand also this one:

https://wiki.openstreetmap.org/wiki/Key:telescope:type

jeisenbe commented:
Most optical telescopes are just tagged as man_made=telescope without
telescope:type. This is unsurprising, because the popular idea of a
telescope is an optical telescope, and these are also the most common.

So that’s why I wanted to render all man_made=telescope with the same icon,
unless they were specifically tagged as radio telescopes.
On Sun, Nov 4, 2018 at 3:34 PM kocio-pl <notifications@github.com> wrote:

> It would be good to expand also this one:
>
> https://wiki.openstreetmap.org/wiki/Key🔭type
> <https://wiki.openstreetmap.org/wiki/Key:telescope:type>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gravitystorm/openstreetmap-carto/pull/3492#issuecomment-435646001>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AoxshPjciWPARqMqdf7LFboqjc4pFaWyks5uroppgaJpZM4YLvuQ>
> .
>

kocio-pl commented on amenity-points.mss:
```diff
@@ -553,6 +553,18 @@
     }
   }
 
+  [feature = 'man_made_telescope']["telescope:type" = 'optical'],["telescope:type" = null] {
```
This should probably look like this:

  [feature = 'man_made_telescope'][telescope:type = 'optical'],
  [feature = 'man_made_telescope'][telescope:type = null] {

because AFAIK this would check if some object has "telescope:type" = null without checking if this is telescope at all, which would lead to strange problems. 

Also I don't think we need adding quotation marks for keys.

kocio-pl replied:
The only key we need quotations is "natural", but this is just because this is a PostgreSQL keyword too by accident.

kocio-pl commented on amenity-points.mss:
```diff
@@ -1867,6 +1879,7 @@
   [feature = 'man_made_silo'][zoom >= 18],
   [feature = 'man_made_communications_tower'][zoom >= 17],
   [feature = 'man_made_telescope']["telescope:type" = 'radio'][zoom >= 16],
+  [feature = 'man_made_telescope'][zoom >= 17],
```
If we use optical vs all the other for icons, I guess name rendering should also check it, not radio vs the rest.

jeisenbe commented:
I thought the quotations were necessary when the key contains a semicolon?

This was also done for tower:type e.g. [ “tower:type” = ‘lighting’ ]

On Tue, Nov 6, 2018 at 7:36 PM kocio-pl <notifications@github.com> wrote:

> *@kocio-pl* commented on this pull request.
> ------------------------------
>
> In amenity-points.mss
> <https://github.com/gravitystorm/openstreetmap-carto/pull/3492#discussion_r231072812>
> :
>
> > @@ -553,6 +553,18 @@
>      }
>    }
>
> +  [feature = 'man_made_telescope']["telescope:type" = 'optical'],["telescope:type" = null] {
>
> The only key we need quotations is "natural", but this is just because
> this is a PostgreSQL keyword too by accident.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gravitystorm/openstreetmap-carto/pull/3492#discussion_r231072812>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AoxshHtaNulHpJRnnqqXgRGwRP3wNeHoks5usWYfgaJpZM4YLvuQ>
> .
>

jeisenbe commented:
I just tried it, and mapnik fails to load, saying:
amenity-points.mss:544:33 Missing closing ] of filter

I did not add or subtract any brackets, I just changed all occurrences of
“telescope:type” to telescope:type in amenity-points.mss and project.mml
On Tue, Nov 6, 2018 at 8:22 PM kocio-pl <notifications@github.com> wrote:

> Please check this just in case, but I doubt it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gravitystorm/openstreetmap-carto/pull/3492#issuecomment-436219807>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AoxshI4RfMJyRBm65ccXZdyBE6rs7uwBks5usXEIgaJpZM4YLvuQ>
> .
>

jeisenbe commented:
I changed the code as suggested, except I did not remove the
telescope:type=radio feature. Now telescope:type I null and optical render
with the done icon and have labels at z17, and only telescope:type = radio
has the dish.

I don’t want to assume that all other types are radio telescopes and render
at z16 with the dish icon. Because it is possible that infrared and x-ray
observatories will be tagged in the future, and gamma telescopes are
already tagged; I believe these should not be rendered the same as radio
telescopes

On Tue, Nov 6, 2018 at 9:11 PM Joseph Eisenberg <joseph.eisenberg@gmail.com>
wrote:

> I just tried it, and mapnik fails to load, saying:
> amenity-points.mss:544:33 Missing closing ] of filter
>
> I did not add or subtract any brackets, I just changed all occurrences of
> “telescope:type” to telescope:type in amenity-points.mss and project.mml
> On Tue, Nov 6, 2018 at 8:22 PM kocio-pl <notifications@github.com> wrote:
>
>> Please check this just in case, but I doubt it.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/gravitystorm/openstreetmap-carto/pull/3492#issuecomment-436219807>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AoxshI4RfMJyRBm65ccXZdyBE6rs7uwBks5usXEIgaJpZM4YLvuQ>
>> .
>>
>

kocio-pl commented:
> I believe these should not be rendered the same as radio telescopes

It's easy to add special icons for them in the future, once we have same sane amount of them (not too much, since they are probably rather rare) and I think it's better to show them at all now.

jeisenbe commented:
I've updated the code as suggested. Here's the rendering of Mauna Kea in Hawaii again.
The NASA Infrared telescope now renders with an icon, but with a dish instead of a dome, unfortunately. However, all the others render correctly. 

(I read up on gamma ray telescopes; they look like dishes. X-ray telescopes are only usable in the high atmosphere or outer space. Radar telescopes are a rare, specialized type of radio telescope. Infrared telescopes are similar to optical telescopes, but only 4 are tagged so far)

Mauna Kea z16
![maunakea-z16](https://user-images.githubusercontent.com/42757252/48132715-d12c7d80-e2d7-11e8-8da3-fffc8de76e1a.png)

kocio-pl commented:
This enhances rendering of telescopes nicely and sane, thanks! 

Some discussion will be needed on Tagging how to make the distinction between touristic view devices and astronomical ones, but there is already some tag with low usage, which might be documented:

https://taginfo.openstreetmap.org/tags/tourism=telescope

geozeisig commented:
I think `tourism=telescope` is the same as `amenity=binoculars`.

[https://taginfo.openstreetmap.org/search?q=amenity%3Dbinoculars](https://taginfo.openstreetmap.org/search?q=amenity%3Dbinoculars)
